File: 2021/day2/Day2.hs
module Day2 (Command, problem1, problem2) where

data Command = Forward | Up | Down

problem1 :: [(Command, Int)] -> Int
problem1 = (\(horizontal, depth) -> horizontal * depth) . foldl control (0, 0)
  where control (horizontal, depth) (Forward, amount) = (horizontal + amount, depth)
        control (horizontal, depth) (Down, amount) = (horizontal, depth + amount)
        control (horizontal, depth) (Up, amount) = (horizontal, depth - amount)

problem2 :: [(Command, Int)] -> Int
problem2 = (\(horizontal, depth, _aim) -> horizontal * depth) . foldl control (0, 0, 0)
  where control (horizontal, depth, aim) (Forward, amount) = (horizontal + amount, depth + (aim * amount), aim)
        control (horizontal, depth, aim) (Down, amount) = (horizontal, depth, aim + amount)
        control (horizontal, depth, aim) (Up, amount) = (horizontal, depth, aim - amount)

File: 2021/day2/day2.ex
defmodule Day2 do
  def problem1() do
    read_file()
    |> Enum.reduce({0, 0}, fn
      {"forward", amount}, {horizontal, depth} ->
        {horizontal + amount, depth}

      {"up", amount}, {horizontal, depth} ->
        {horizontal, depth - amount}

      {"down", amount}, {horizontal, depth} ->
        {horizontal, depth + amount}
    end)
    |> then(fn {horizontal, depth} -> horizontal * depth end)
  end

  def problem2() do
    read_file()
    |> Enum.reduce({0, 0, 0}, fn
      {"forward", amount}, {horizontal, depth, aim} ->
        {horizontal + amount, depth + aim * amount, aim}

      {"up", amount}, {horizontal, depth, aim} ->
        {horizontal, depth, aim - amount}

      {"down", amount}, {horizontal, depth, aim} ->
        {horizontal, depth, aim + amount}
    end)
    |> then(fn {horizontal, depth, _aim} -> horizontal * depth end)
  end

  defp read_file() do
    "input.txt"
    |> File.read!()
    |> String.split("\n")
    |> Enum.reject(&(&1 == ""))
    |> Enum.map(&String.split(&1, " "))
    |> Enum.map(fn [command, amount] -> {command, String.to_integer(amount)} end)
  end
end
